src/transform/src/predicate_propagation.rs

```diff
@@ -0,0 +1,837 @@
+                } => {
+                    let input_knowledge = self.action(input, let_knowledge)?;
+                    let structured = PredicateStructure::new(&input_knowledge);
+                    for expr in exprs {
+                        optimize(expr, &structured);
+                    }
+                    input_knowledge
+                }
+                MirRelationExpr::Filter { input, predicates } => {
+                    let input_knowledge = self.action(input, let_knowledge)?;
+                    let structured = PredicateStructure::new(&input_knowledge);
+                    let mut output_knowledge = input_knowledge.clone();
+
+                    for predicate in predicates.iter_mut() {
+                        optimize(predicate, &structured);
+                        append_constraint(&mut output_knowledge, predicate.clone());
+                    }
+                    output_knowledge
+                }
+                MirRelationExpr::Join {
+                    inputs,
+                    equivalences,
+                    ..
+                } => {
+                    let input_mapper = JoinInputMapper::new(inputs);
+
+                    let mut knowledge = Vec::new();
+
+                    // Collect input knowledge, but update column references.
+                    for (input_idx, input) in inputs.iter_mut().enumerate() {
+                        for predicate in self.action(input, let_knowledge)? {
+                            knowledge.push(predicate.map_to_global(&input_mapper, input_idx));
+                        }
+                    }
+
+                    let structured = PredicateStructure::new(&knowledge);
+                    let mut new_predicates = Vec::new();
+
+                    for equivalence in equivalences.iter_mut() {
+                        for expr in equivalence.iter_mut() {
+                            optimize(expr, &structured);
+                        }
+
+                        new_predicates.push(Constraint::Equivalence(equivalence.clone()));
+                    }
+
+                    knowledge.extend(new_predicates);
+                    knowledge
+                }
+                MirRelationExpr::Reduce {
+                    input,
+                    group_key,
+                    aggregates,
+                    ..
+                } => {
+                    let input_type = input.typ();
+                    let input_knowledge = self.action(input, let_knowledge)?;
+                    let structured = PredicateStructure::new(&input_knowledge);
+                    for key in group_key.iter_mut() {
+                        optimize(key, &structured);
+                    }
+                    for aggregate in aggregates.iter_mut() {
+                        optimize(&mut aggregate.expr, &structured);
+                    }
+
+                    // List of predicates we will return.
+                    let mut predicates = Vec::new();
```

Comment:
nit: rename this and other variables `predicates` of type `Vec<Constraint>` to `constraints`.